src/dispatch/stage.rs

```diff
@@ -242,20 +242,42 @@ impl<'a> StagesBuilder<'a> {
         let new_writes = new_writes.into_iter();
 
         let num_groups = ids[stage].len();
+        let mut dep_conflict = false;
 
         let conflict = (0..num_groups)
             .filter(|&group| {
                 let read_and_write = writes[stage][group]
                     .iter()
                     .chain(reads[stage][group].iter());
 
-                check_intersection(new_writes.clone(), read_and_write) ||
-                check_intersection(new_reads.clone(), writes[stage][group].iter()) ||
-                check_intersection(new_dep.iter(), ids[stage][group].iter())
+                let inters = check_intersection(new_writes.clone(), read_and_write) ||
+                check_intersection(new_reads.clone(), writes[stage][group].iter());
+
+                if !inters {
+                    if check_intersection(new_dep.iter(), ids[stage][group].iter()) {
+                        dep_conflict = true;
+
+                        true
+                    } else {
+                        false
+                    }
+                } else {
+                    true
+                }
+
             })
             .fold(Conflict::None, Conflict::add);
 
-        conflict
+        // If there is a dependency in the dependency list
+        // which was not in a previous or this stage,
+        // return `Multiple` conflict.
+
+        match conflict {
+            Conflict::None if !new_dep.is_empty() => Conflict::Multiple,
+            Conflict::Single(_) if (dep_conflict && new_dep.len() > 1) || !new_dep.is_empty() => Conflict::Multiple,
```

Comment:
I am not sure what this is supposed to be doing, but it seems wrong: if `new_dep.len() > 1` is true then `!new_dep.is_empty()` is also true, which means that `(dep_conflict && new_dep.len() > 1)` doesn't do anything.

Reply:
Also it's kinda annoying that there is `!new_dep.is_empty()` in multiple guards, but I couldn't come up with better formulation.

Reply:
Sounds like a lack of expressiveness in `Conflict::Multiple`

Reply:
Yeah. Maybe I should rename the variant?

Reply:
@WaDelma Here's how this works:

If no conflict has been found (`Conflict::None`), there is no intersection reads / writes / dependencies. Now, just because there is no intersection, doesn't mean we can insert the system right here.

Here's why: There can still be dependencies to systems, which are e.g. in the next stage. That means, I have to check if the dependency list is empty (note that I remove dependencies once I iterated over them).

The other case where we cannot insert the system in this stage is if there is

i) a single conflict and
ii) that conflict is not the single dependency in the list.

Reply:
If you want a more general overview, there's a short comment at the beginning of the file.

Reply:
Does the latest commit fix it? As there is still the same logic problem with `(dep_conflict && new_dep.len() > 1) || unmet_dependency`.